File: vignettes/23_skills/python/23_skills.py
"""
Skills System (Python)

This sample demonstrates the Agent Skills system: reusable, discoverable
instruction sets that extend agent capabilities via progressive disclosure.
It mirrors the Julia vignette 23_skills.

Prerequisites:
  - Ollama running locally with `qwen3:8b` pulled
  - pip install agent-framework-ollama
"""

import asyncio
import json

from textwrap import dedent
from typing import Any

from agent_framework import Agent, Skill, SkillResource, SkillsProvider
from agent_framework.ollama import OllamaChatClient


# ── Parsing SKILL.md files ───────────────────────────────────────────────────

# File-based skills are discovered from SKILL.md files on disk.
# Each skill lives in its own directory:
#
#   skills/
#   ├── data_analysis/
#   │   ├── SKILL.md
#   │   └── schema.json
#   └── code_review/
#       ├── SKILL.md
#       └── style_guide.md
#
# The SKILL.md format uses YAML frontmatter:
#
#   ---
#   name: DataAnalysis
#   description: Skill for analyzing tabular data
#   version: 1.0.0
#   tags: [data, analysis, csv]
#   ---
#
#   # Data Analysis Instructions
#
#   When asked to analyze data:
#   1. First understand the structure ...


# ── Code-Defined Skills ──────────────────────────────────────────────────────

# 1. Create a skill with static resources (inline content).
unit_converter_skill = Skill(
    name="unit-converter",
    description="Convert between common units using a conversion factor",
    content=dedent("""\
        Use this skill when the user asks to convert between units.

        1. Review the conversion-tables resource to find the factor for the
           requested conversion.
        2. Check the conversion-policy resource for rounding and formatting rules.
        3. Use the convert script, passing the value and factor from the table.
    """),
    resources=[
        SkillResource(
            name="conversion-tables",
            content=dedent("""\
                # Conversion Tables

                Formula: **result = value × factor**

                | From        | To          | Factor   |
                |-------------|-------------|----------|
                | miles       | kilometers  | 1.60934  |
                | kilometers  | miles       | 0.621371 |
                | pounds      | kilograms   | 0.453592 |
                | kilograms   | pounds      | 2.20462  |
            """),
        ),
    ],
)


# 2. Dynamic resource — callable function via @skill.resource decorator.
@unit_converter_skill.resource(
    name="conversion-policy",
    description="Current conversion formatting and rounding policy",
)
def conversion_policy(**kwargs: Any) -> Any:
    """Return the current conversion policy (computed at runtime)."""
    precision = kwargs.get("precision", 4)
    return dedent(f"""\
        # Conversion Policy

        **Decimal places:** {precision}
        **Format:** Always show both the original and converted values with units
    """)


# 3. Dynamic script — in-process callable.
@unit_converter_skill.script(
    name="convert",
    description="Convert a value: result = value × factor",
)
def convert_units(value: float, factor: float, **kwargs: Any) -> str:
    """Convert a value using a multiplication factor."""
    precision = kwargs.get("precision", 4)
    result = round(value * factor, precision)
    return json.dumps({"value": value, "factor": factor, "result": result})


# ── Using SkillsProvider with an Agent ───────────────────────────────────────

async def code_defined_skills_example() -> None:
    """Demonstrate code-defined skills with an agent."""
    print("=== Code-Defined Skills ===\n")

    client = OllamaChatClient(
        host="http://localhost:11434",
        model="qwen3:8b",
    )

    # Pass skills directly to the provider — no files needed.
    async with Agent(
        client=client,
        instructions="You are a helpful assistant that can convert units.",
        context_providers=[SkillsProvider(skills=[unit_converter_skill])],
    ) as agent:
        response = await agent.run(
            "How many kilometers is a marathon (26.2 miles)?",
        )
        print(f"Agent: {response.text}\n")


# ── File-Based Skills ────────────────────────────────────────────────────────

async def file_based_skills_example() -> None:
    """Demonstrate file-based skills discovered from a directory."""
    print("=== File-Based Skills ===\n")

    client = OllamaChatClient(
        host="http://localhost:11434",
        model="qwen3:8b",
    )

    # Discover skills from a directory of SKILL.md files.
    skills_provider = SkillsProvider(skill_paths="./skills")

    async with Agent(
        client=client,
        instructions="You have access to various skills. Use them when appropriate.",
        context_providers=[skills_provider],
    ) as agent:
        response = await agent.run("Analyze this data: name,age\nAlice,30\nBob,25")
        print(f"Agent: {response.text}\n")


# ── Mixed Skills (Code + File) ───────────────────────────────────────────────

async def mixed_skills_example() -> None:
    """Combine code-defined and file-based skills in a single agent."""
    print("=== Mixed Skills ===\n")

    client = OllamaChatClient(
        host="http://localhost:11434",
        model="qwen3:8b",
    )

    # Combine code skills with file-based discovery.
    skills_provider = SkillsProvider(
        skill_paths="./skills",
        skills=[unit_converter_skill],
    )

    async with Agent(
        client=client,
        instructions="You are a helpful assistant with access to various skills.",
        context_providers=[skills_provider],
    ) as agent:
        response = await agent.run(
            "How many kilometers is a marathon (26.2 miles)? "
            "And how many pounds is 75 kilograms?"
        )
        print(f"Agent: {response.text}\n")


# ── Main ─────────────────────────────────────────────────────────────────────

async def main() -> None:
    # Code-defined skills (no files needed).
    await code_defined_skills_example()

    # File-based skills — uncomment if you have a skills/ directory:
    # await file_based_skills_example()

    # Mixed skills — uncomment if you have a skills/ directory:
    # await mixed_skills_example()


if __name__ == "__main__":
    asyncio.run(main())

File: vignettes/23_skills/csharp/23_skills.cs
// Skills System (C#)
//
// This sample demonstrates the Agent Skills system: reusable, discoverable
// instruction sets that extend agent capabilities via progressive disclosure.
// It mirrors the Julia vignette 23_skills.
//
// Prerequisites:
//   - Ollama running locally with qwen3:8b pulled
//   - dotnet restore
//   - NuGet: Microsoft.Agents.AI, OllamaSharp

using System.Text.Json;
using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OllamaSharp;

var endpoint = Environment.GetEnvironmentVariable("OLLAMA_ENDPOINT")
    ?? "http://localhost:11434";
var modelName = Environment.GetEnvironmentVariable("OLLAMA_MODEL_NAME")
    ?? "qwen3:8b";

// ── Code-Defined Skills ─────────────────────────────────────────────────────

// Define a skill entirely in code using AgentInlineSkill.
// No SKILL.md files needed — skills, resources, and scripts are programmatic.

Console.WriteLine("=== Code-Defined Skills ===\n");

var unitConverterSkill = new AgentInlineSkill(
    name: "unit-converter",
    description: "Convert between common units using a multiplication factor",
    instructions: """
        Use this skill when the user asks to convert between units.

        1. Review the conversion-tables resource to find the factor.
        2. Check the conversion-policy resource for rounding rules.
        3. Use the convert script, passing the value and factor.
        """)
    // Static resource: inline content.
    .AddResource(
        "conversion-tables",
        """
        # Conversion Tables

        Formula: **result = value × factor**

        | From        | To          | Factor   |
        |-------------|-------------|----------|
        | miles       | kilometers  | 1.60934  |
        | kilometers  | miles       | 0.621371 |
        | pounds      | kilograms   | 0.453592 |
        | kilograms   | pounds      | 2.20462  |
        """)
    // Dynamic resource: computed at runtime.
    .AddResource("conversion-policy", () =>
    {
        const int Precision = 4;
        return $"""
            # Conversion Policy

            **Decimal places:** {Precision}
            **Format:** Always show both the original and converted values with units
            **Generated at:** {DateTime.UtcNow:O}
            """;
    })
    // Code script: executable delegate the agent can invoke.
    .AddScript("convert", (double value, double factor) =>
    {
        double result = Math.Round(value * factor, 4);
        return JsonSerializer.Serialize(new { value, factor, result });
    });

// Create a SkillsProvider with the code-defined skill.
var codeSkillsProvider = new AgentSkillsProvider(unitConverterSkill);

AIAgent codeAgent = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(new ChatClientAgentOptions
    {
        Name = "UnitConverterAgent",
        ChatOptions = new()
        {
            Instructions = "You are a helpful assistant that can convert units.",
        },
        AIContextProviders = [codeSkillsProvider],
    });

Console.WriteLine("Converting units with code-defined skills");
Console.WriteLine(new string('-', 60));

AgentResponse codeResponse = await codeAgent.RunAsync(
    "How many kilometers is a marathon (26.2 miles)?");
Console.WriteLine($"Agent: {codeResponse.Text}\n");

// ── File-Based Skills ───────────────────────────────────────────────────────

// Discover skills from SKILL.md files in a directory tree.
// Each skill lives in its own folder:
//
//   skills/
//   ├── data_analysis/
//   │   ├── SKILL.md
//   │   └── schema.json
//   └── code_review/
//       ├── SKILL.md
//       └── style_guide.md

Console.WriteLine("=== File-Based Skills ===\n");

// Uncomment when you have a skills/ directory:
// var fileSkillsProvider = new AgentSkillsProvider(
//     Path.Combine(AppContext.BaseDirectory, "skills"));
//
// AIAgent fileAgent = new OllamaApiClient(new Uri(endpoint), modelName)
//     .AsAIAgent(new ChatClientAgentOptions
//     {
//         Name = "SkillfulAgent",
//         ChatOptions = new()
//         {
//             Instructions = "You have access to various skills. Use them when appropriate.",
//         },
//         AIContextProviders = [fileSkillsProvider],
//     });
//
// AgentResponse fileResponse = await fileAgent.RunAsync(
//     "Analyze this data: name,age\nAlice,30\nBob,25");
// Console.WriteLine($"Agent: {fileResponse.Text}\n");

Console.WriteLine("(Skipped — no skills/ directory present)\n");

// ── Mixed Skills (Code + File) ──────────────────────────────────────────────

// Use AgentSkillsProviderBuilder to combine multiple skill sources.

Console.WriteLine("=== Mixed Skills ===\n");

// Uncomment when you have a skills/ directory:
// var mixedProvider = new AgentSkillsProviderBuilder()
//     .UseFileSkill(Path.Combine(AppContext.BaseDirectory, "skills"))
//     .UseSkill(unitConverterSkill)
//     .Build();
//
// AIAgent mixedAgent = new OllamaApiClient(new Uri(endpoint), modelName)
//     .AsAIAgent(new ChatClientAgentOptions
//     {
//         Name = "MultiSkillAgent",
//         ChatOptions = new()
//         {
//             Instructions = "You are a helpful assistant with access to various skills.",
//         },
//         AIContextProviders = [mixedProvider],
//     });
//
// AgentResponse mixedResponse = await mixedAgent.RunAsync(
//     "How many kilometers is 26.2 miles? And how many pounds is 75 kilograms?");
// Console.WriteLine($"Agent: {mixedResponse.Text}\n");

Console.WriteLine("(Skipped — no skills/ directory present)\n");

Console.WriteLine("Done.");
